urlparse4/cgurl.pyx

```diff
@@ -303,6 +325,27 @@ class ParsedResultNamedTuple(tuple):
         if not scheme and input_scheme:
             scheme = input_scheme.encode('utf-8')
 
+        # encode based on the encoding input
+        if canonicalize and canonicalize_encoding != 'utf-8':
+            if query:
+                try:
+                    query = query.decode().encode(canonicalize_encoding)
```

Comment:
I don't quite understand this line - is query first bytes, which are decoded using utf-8 and then encoded to specified encoding?

Reply:
yes, if it the encoding is specified as something other than 'utf-8', the current query (or ref) is byte-type of utf-8. Therefore, I had to convert it back to str type of utf-8 before i can convert it to the byte type of the specified encoding!